File: advent2017/extras/cxx/src/days/day19.cpp
#include <chrono>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "b4231dede8cc9f00c1dcdf6fe60b2c5cc33278020531f4a05af462099063171a";
const std::string kExpectedPart1 = "DTOUFARJQ";
const std::string kExpectedPart2 = "16642";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{"advent2017/Day19/d19_input.txt", "Day19/d19_input.txt", "../Day19/d19_input.txt", "../../Day19/d19_input.txt"};
    for (const auto& c : cands) if (std::filesystem::exists(c)) return c;
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::vector<std::string> parse_grid(const std::string& raw) {
    std::vector<std::string> lines;
    std::istringstream in(raw);
    std::string line;
    std::size_t w = 0;
    while (std::getline(in, line)) {
        lines.push_back(line);
        w = std::max(w, line.size());
    }
    for (auto& l : lines) if (l.size() < w) l += std::string(w - l.size(), ' ');
    return lines;
}

char at(const std::vector<std::string>& g, int r, int c) {
    if (r < 0 || c < 0 || r >= static_cast<int>(g.size()) || c >= static_cast<int>(g[0].size())) return ' ';
    return g[r][c];
}

std::pair<std::string, long long> solve(const std::string& raw) {
    const auto g = parse_grid(raw);
    int r = 0;
    int c = static_cast<int>(g[0].find('|'));
    int dr = 1, dc = 0;
    std::string letters;
    long long steps = 0;

    while (true) {
        const char ch = at(g, r, c);
        if (ch == ' ') break;
        if (ch >= 'A' && ch <= 'Z') letters.push_back(ch);
        else if (ch == '+') {
            if (dr != 0) {
                if (at(g, r, c - 1) != ' ') { dr = 0; dc = -1; }
                else if (at(g, r, c + 1) != ' ') { dr = 0; dc = 1; }
            } else {
                if (at(g, r - 1, c) != ' ') { dr = -1; dc = 0; }
                else if (at(g, r + 1, c) != ' ') { dr = 1; dc = 0; }
            }
        }
        r += dr;
        c += dc;
        steps++;
    }

    return {letters, steps};
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");

    auto t0 = std::chrono::steady_clock::now();
    const auto [p1, p2] = solve(raw);
    const std::string ans = (part == 1) ? p1 : std::to_string(p2);
    const std::string expected = (part == 1) ? kExpectedPart1 : kExpectedPart2;
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    const double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=19 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day19.rs
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "b4231dede8cc9f00c1dcdf6fe60b2c5cc33278020531f4a05af462099063171a";
const EXPECTED_PART1: &str = "DTOUFARJQ";
const EXPECTED_PART2: &str = "16642";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() { return provided.to_string(); }
    let cands = ["advent2017/Day19/d19_input.txt", "Day19/d19_input.txt", "../Day19/d19_input.txt", "../../Day19/d19_input.txt"];
    for c in cands { if Path::new(c).exists() { return c.to_string(); } }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new(); h.update(&b); hex::encode(h.finalize())
}

fn parse_grid(raw: &str) -> Vec<Vec<u8>> {
    let lines = raw.trim_end_matches('\n').lines().collect::<Vec<&str>>();
    let w = lines.iter().map(|l| l.len()).max().unwrap_or(0);
    lines
        .iter()
        .map(|l| {
            let mut row = vec![b' '; w];
            row[..l.len()].copy_from_slice(l.as_bytes());
            row
        })
        .collect::<Vec<Vec<u8>>>()
}

fn at(g: &[Vec<u8>], r: i32, c: i32) -> u8 {
    if r < 0 || c < 0 || (r as usize) >= g.len() || (c as usize) >= g[0].len() { b' ' }
    else { g[r as usize][c as usize] }
}

fn solve(raw: &str) -> (String, i64) {
    let g = parse_grid(raw);
    let mut r: i32 = 0;
    let mut c: i32 = g[0].iter().position(|&x| x == b'|').unwrap() as i32;
    let mut dr: i32 = 1;
    let mut dc: i32 = 0;
    let mut letters: Vec<u8> = Vec::new();
    let mut steps: i64 = 0;

    loop {
        let ch = at(&g, r, c);
        if ch == b' ' { break; }
        if ch.is_ascii_uppercase() {
            letters.push(ch);
        } else if ch == b'+' {
            if dr != 0 {
                if at(&g, r, c - 1) != b' ' { dr = 0; dc = -1; }
                else if at(&g, r, c + 1) != b' ' { dr = 0; dc = 1; }
            } else {
                if at(&g, r - 1, c) != b' ' { dr = -1; dc = 0; }
                else if at(&g, r + 1, c) != b' ' { dr = 1; dc = 0; }
            }
        }
        r += dr;
        c += dc;
        steps += 1;
    }

    (String::from_utf8(letters).unwrap(), steps)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => { part = args[i + 1].parse().unwrap(); i += 2; }
            "--input" => { input = args[i + 1].clone(); i += 2; }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 { panic!("--part 1|2 required"); }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA { panic!("checksum mismatch"); }
    let raw = fs::read_to_string(&in_path).expect("read_to_string");

    let t0 = Instant::now();
    let (p1, p2) = solve(&raw);
    let ans = if part == 1 { p1 } else { p2.to_string() };
    let expected = if part == 1 { EXPECTED_PART1 } else { EXPECTED_PART2 };
    if ans != expected { panic!("answer mismatch"); }

    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=19 part={} runtime_ms={:.3}", part, ms);
}
